### PR TITLE
JAMES-316 API and contract for EmailAddressContact, in memory implementation

### PR DESCRIPTION
**Pull request by tieugiang91**

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+trait EmailAddressContactSearchEngine {
+  def index(accountId: AccountId, contact: EmailAddressContact): Publisher[Unit]
+
+  def autoComplete(accountId: AccountId, part: String): Publisher[EmailAddressContact]
+
+}
+
+class InMemoryEmailAddressContactSearchEngine extends EmailAddressContactSearchEngine {
+  var emailList: Map[AccountId, Set[EmailAddressContact]] = Map()
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala:23]: Guava multimaps should be the way to go IMO (Use Synchronized multimap)

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/model/EmailAddressContactSearchEngineTest.scala [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.linagora.tmail.james.jmap.model
+
+import org.junit.jupiter.api.BeforeEach
+
+class EmailAddressContactSearchEngineTest extends EmailAddressContactSearchEngineContract {
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/model/EmailAddressContactSearchEngineTest.scala:5]: MemoryEmailAddressContactSearchEngineTest

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala [hunk 3]
```diff
@@ -20,24 +23,25 @@ trait EmailAddressContactSearchEngine {
...
-    var addressList: Set[EmailAddressContact] = Set()
-    if (emailList.contains(accountId)) {
-      addressList = emailList.apply(accountId)
+    var addressList: Iterable[EmailAddressContact] = Set()
+    if (emailList.containsKey(accountId)) {
+      addressList = emailList.get(accountId).asScala
     }
     val fullList: SFlux[EmailAddressContact] = SFlux.fromIterable(addressList)
-    fullList.filter(e => e.address.contains(part))
+    fullList.filter(_.address.contains(part))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala]: 1. Don't use `var`
2. get rid of the `if`
3. Wrap everything in a SMono. Just like the method above.

File: tmail-backend/pom.xml [hunk 4]
```diff
@@ -290,6 +290,11 @@
                 <artifactId>bcpg-jdk15on</artifactId>
                 <version>1.69</version>
             </dependency>
+            <dependency>
+                <groupId>com.google.guava</groupId>
+                <artifactId>guava</artifactId>
+                <version>31.0.1</version>
```
Comment (chibenwa) [inline tmail-backend/pom.xml:296]: No please do not override the guavva version: we gonna get problems!

Why do you need guava 30.0.1 especially?

**Review by Arsnael — COMMENTED** [review]
Read it

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraKeystoreGetMethodContract.scala [hunk 5]
```diff
@@ -1,10 +1,10 @@
 package com.linagora.tmail.james.common
 
 import java.nio.charset.StandardCharsets
-
 import com.linagora.tmail.james.common.LinagoraKeystoreGetMethodContract.{PGP_KEY_ARMORED, PGP_KEY_ARMORED2, PGP_KEY_ID, PGP_KEY_ID2}
 import io.netty.handler.codec.http.HttpHeaderNames.ACCEPT
 import io.restassured.RestAssured.{`given`, requestSpecification}
+import io.restassured.filter.log.LogDetail
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraKeystoreGetMethodContract.scala:7]: Are the changes in the imports needed?

File: tmail-backend/jmap/extensions/pom.xml [hunk 6]
```diff
@@ -58,6 +58,10 @@
             <artifactId>mockito-core</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/pom.xml:62]: We tend to sort dependencies alphabetically by groupId then by artifactId

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/model/MemoryEmailAddressContactSearchEngineTest.scala [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package com.linagora.tmail.james.jmap.model
+
+import org.junit.jupiter.api.BeforeEach
+
+class MemoryEmailAddressContactSearchEngineTest extends EmailAddressContactSearchEngineContract {
+  var inMemoryEmailAddressContactSearchEngine: InMemoryEmailAddressContactSearchEngine = _
+  override def testee: EmailAddressContactSearchEngine = inMemoryEmailAddressContactSearchEngine
+  @BeforeEach
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/model/MemoryEmailAddressContactSearchEngineTest.scala:8]: Please line breaks to ease readability.

File: tmail-backend/jmap/extensions/pom.xml (hunk 6 above)
Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/pom.xml:62]: Let use ```mvn sortpom:sort``` for quickly sort this.

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraKeystoreGetMethodContract.scala (hunk 5 above)
Comment (quantranhong1999) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraKeystoreGetMethodContract.scala:7]: Use Ctrl+Alt+O to optimize imports.

File: tmail-backend/pom.xml (hunk 4 above)
Comment (tieugiang91) [reply tmail-backend/pom.xml]: I just stupidly did that, it has been fixed

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala [hunk 8]
```diff
@@ -20,24 +23,25 @@ trait EmailAddressContactSearchEngine {
...
-    }
-    mailListPerAcc = mailListPerAcc + address
-    emailList = emailList + (accountId -> mailListPerAcc)
-    SFlux.fromIterable(emailList).`then`()
+    SMono.fromCallable(() => {
+      if (emailList.containsKey(accountId)) {
+        emailList.get(accountId).add(address)
+      } else {
+        emailList.put(accountId, address)
+      }
```
[the remark this answers is not in the document]
Comment (tieugiang91) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala:31]: There's no difference, fixed it already.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+  }
+
+  override def autoComplete(accountId: AccountId, part: String): Publisher[EmailAddressContact] = {
+    var addressList: Iterable[EmailAddressContact] = Set()
+    if (emailList.containsKey(accountId)) {
+      addressList = emailList.get(accountId).asScala
+    }
+    val fullList: SFlux[EmailAddressContact] = SFlux.fromIterable(addressList)
+    fullList.filter(_.address.contains(part))
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala:42]: I think you can put everything here easily in a SFlux?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala (hunk 3 above)
Comment (tieugiang91) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala]: Seems I used SFlux only without using SMono.just. Is it okie?

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraKeystoreGetMethodContract.scala [hunk 10]
```diff
@@ -1,21 +1,21 @@
...
 import org.apache.james.jmap.core.ResponseObject.SESSION_STATE
 import org.apache.james.jmap.core.UuidState.INSTANCE
 import org.apache.james.jmap.http.UserCredential
-import org.apache.james.jmap.rfc8621.contract.Fixture.{ACCEPT_RFC8621_VERSION_HEADER, ACCOUNT_ID, BOB, BOB_PASSWORD, DOMAIN, authScheme, baseRequestSpecBuilder}
+import org.apache.james.jmap.rfc8621.contract.Fixture._
 import org.apache.james.utils.DataProbeImpl
 import org.junit.jupiter.api.{BeforeEach, Test}
 
+import java.nio.charset.StandardCharsets
+
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraKeystoreGetMethodContract.scala]: Can we avoid changes in unrelated files?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala [hunk 11]
```diff
@@ -34,11 +34,6 @@ class InMemoryEmailAddressContactSearchEngine extends EmailAddressContactSearchE
   }
 
   override def autoComplete(accountId: AccountId, part: String): Publisher[EmailAddressContact] = {
-    var addressList: Iterable[EmailAddressContact] = Set()
-    if (emailList.containsKey(accountId)) {
-      addressList = emailList.get(accountId).asScala
-    }
-    val fullList: SFlux[EmailAddressContact] = SFlux.fromIterable(addressList)
-    fullList.filter(_.address.contains(part))
+    SFlux.fromIterable(emailList.get(accountId).asScala).filter(_.address.contains(part))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala:37]: We can filter on the flux.

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/model/MemoryEmailAddressContactSearchEngineTest.scala [hunk 12]
```diff
@@ -5,6 +5,7 @@ import org.junit.jupiter.api.BeforeEach
 class MemoryEmailAddressContactSearchEngineTest extends EmailAddressContactSearchEngineContract {
   var inMemoryEmailAddressContactSearchEngine: InMemoryEmailAddressContactSearchEngine = _
   override def testee: EmailAddressContactSearchEngine = inMemoryEmailAddressContactSearchEngine
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/model/MemoryEmailAddressContactSearchEngineTest.scala]:
```suggestion
  var inMemoryEmailAddressContactSearchEngine: InMemoryEmailAddressContactSearchEngine = _
  
  override def testee: EmailAddressContactSearchEngine = inMemoryEmailAddressContactSearchEngine
```

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraKeystoreGetMethodContract.scala (hunk 10 above)
Comment (tieugiang91) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraKeystoreGetMethodContract.scala]: I'll revert it to the previous commit on Sep 28th, which is the latest change not by me.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala (hunk 11 above)
Comment (tieugiang91) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala:37]: May you clarify this comment please?

**Review by chibenwa — APPROVED** [review]
Good to me.

Let's work on the ElasticSearch version now <3

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Just a remark for next PRs the commit comments like Fix comments on PR... are not really relevant. If you fix comments in a PR, do a fixup! [commit message] 

So we know what to squash together before merging to the master branch. We don't want to have a git history mess in our master branch, it needs to stay clear and easy for the guy that will merge your work (and yourself as well believe me^^)

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  def index(accountId: AccountId, contact: EmailAddressContact): Publisher[Unit]
+
+  def autoComplete(accountId: AccountId, part: String): Publisher[EmailAddressContact]
+
+}
+
+class InMemoryEmailAddressContactSearchEngine extends EmailAddressContactSearchEngine {
+  val emailList: Multimap[AccountId, EmailAddressContact] = Multimaps.synchronizedSetMultimap(HashMultimap.create())
+
+  override def index(accountId: AccountId, address: EmailAddressContact): Publisher[Unit] = {
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala:25]: Benoit always remind me to remove unnecessary bracket `{}` when the method is short

Comment (tieugiang91) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala]: Done fix, looks much better, thanks, Tung :D

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by quantranhong1999** [comment]
@tieugiang91 I think you should clean git commit history a little here.
